### DISCUSSION TITLE
Middleware to track timestamp of last updated

### DISCUSSION THREAD

**Discussion by xypnox** [pmndrs/zustand#1196] [Q&A]
I am writing a middleware to track the timestamp of the last time the zustand store was updated with new state.

With inspiration from [zustand-middleware-computed-state](https://github.com/cmlarsen/zustand-middleware-computed-state/blob/main/src/index.ts), I have managed to write `metaState.ts`:

```ts
import { GetState, SetState, State, StateCreator, StoreApi } from 'zustand'

export interface ZustandStoreMeta {
  meta: {
    lastUpdated: number
    updated: boolean
  }
}

export const baseMetaState  = {
  meta: {
    lastUpdated: 0,
    updated: false
  }
}

export const metaState =
  <S extends State & ZustandStoreMeta>(create: StateCreator<S>) =>
  (set: SetState<S>, get: GetState<S>, api: StoreApi<S>): S => {
    const setWithUpdateTimestamp: SetState<S> = (update, replace) => {
      set((state) => {
      const updated = typeof update === 'object' ? update : update(state)
      const mState: S = {
          ...updated,
          meta: {
            ...state.meta,
            updated: true,
            lastUpdated: Date.now()
          }
        }
        return mState
      }, replace)
    }

    api.setState = setWithUpdateTimestamp
    const state = create(setWithUpdateTimestamp, get, api)
    return {
      ...state,
      ...baseMetaState
    }
  }
```

However, this concoction throws error with type of mState:

```
ERROR in metaState.ts:23:15
TS2322: Type '(Pick<S, K1> & { meta: { updated: true; lastUpdated: number; }; }) | (Pick<S, K2> & { meta: { updated: true; lastUpdated: number; }; }) | (Pick<S, K3> & { ...; }) | (Pick<...> & { ...; })' is not assignable to type 'S'.
  '(Pick<S, K1> & { meta: { updated: true; lastUpdated: number; }; }) | (Pick<S, K2> & { meta: { updated: true; lastUpdated: number; }; }) | (Pick<S, K3> & { ...; }) | (Pick<...> & { ...; })' is assignable to the constraint of type 'S', but 'S' could be instantiated with a different subtype of constraint 'object & ZustandStoreMeta'.
    Type 'Pick<S, K1> & { meta: { updated: true; lastUpdated: number; }; }' is not assignable to type 'S'.
      'Pick<S, K1> & { meta: { updated: true; lastUpdated: number; }; }' is assignable to the constraint of type 'S', but 'S' could be instantiated with a different subtype of constraint 'object & ZustandStoreMeta'.
    21 |       set((state) => {
    22 |         const updated = typeof update === 'object' ? update : update(state)
  > 23 |         const mState: S = {
       |               ^^^^^^
    24 |           ...updated,
    25 |           meta: {
    26 |             ...state.meta,
```

What does the specific error mean and how can I resolve it?

**Clarification:**

Removing the `S` type from mState throws the following error:

```
ERROR in metaState.ts:21:11
TS2345: Argument of type '(state: S) => (Pick<S, K1> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<S, K2> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<...> & { ...; }) | (Pick<...> & { ...; })' is not assignable to parameter of type 'PartialState<S, K1 | K2 | K3 | K4 | "meta", K1 | K2 | K3 | K4 | "meta", K1 | K2 | K3 | K4 | "meta", K1 | K2 | K3 | K4 | "meta">'.
  Type '(state: S) => (Pick<S, K1> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<S, K2> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<...> & { ...; }) | (Pick<...> & { ...; })' is not assignable to type '(state: S) => S | Pick<S, K1 | K2 | K3 | K4 | "meta">'.
    Type '(Pick<S, K1> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<S, K2> & { meta: { updated: boolean; lastUpdated: number; }; }) | (Pick<S, K3> & { ...; }) | (Pick<...> & { ...; })' is not assignable to type 'S | Pick<S, K1 | K2 | K3 | K4 | "meta">'.
      Type 'Pick<S, K1> & { meta: { updated: boolean; lastUpdated: number; }; }' is not assignable to type 'S | Pick<S, K1 | K2 | K3 | K4 | "meta">'.
        Type 'Pick<S, K1> & { meta: { updated: boolean; lastUpdated: number; }; }' is not assignable to type 'Pick<S, K1 | K2 | K3 | K4 | "meta">'.
    19 |   (set: SetState<S>, get: GetState<S>, api: StoreApi<S>): S => {
    20 |     const setWithUpdateTimestamp: SetState<S> = (update, replace) => {
  > 21 |       set((state) => {
       |           ^^^^^^^^^^^^
    22 |         const updated = typeof update === 'object' ? update : update(state)
    23 |         const mState = {
    24 |           ...updated,
```

**Comment by dai-shi** [comment]
https://tsplay.dev/W4x4KW
Just removing `: S`? (but there's another issue I don't know how to solve.)

**Reply by xypnox** [reply]
I have tried removing the S too. I added the S to narrow down the error as it should return the State of type `S`.

I have updated the description with the different error, if it helps.

**Comment by dai-shi** [comment]
@devanshj can help??

**Comment by devanshj** [comment]
The type of `metaState` is a bit tricky, there are two options:
1. The simpler one: https://tsplay.dev/Nng4dm. Here the user will have to pass `WithMeta<BearState>` instead of just `BearState`
2. The more correct one: https://tsplay.dev/WkKjJW. `metaState` kind of becomes the `create`, hence `metaState` is curried (instead of `create`) and then user passes `BearState` to it instead of `create`, here the pro is that user doesn't have to pass `WithMeta<BearState>` but just `BearState`. This approach is similiar to how `redux` and `combine` are typed, because they too "create state" and when using them `create` is not curried as the [typescript doc mentions it too](https://github.com/pmndrs/zustand/blob/main/docs/typescript.md#basic-usage)...

    > Also note that we're not using the curried version when using combine because combine "creates" the state. When using a middleware that creates the state, it's not necessary to use the curried version because the state now can be inferred. Another middleware that creates state is redux. So when using combine, redux or any other custom middleware that creates the state, it's not recommended to use the curried version.

I know the types a bit complex, for reference one could see the ["Middleware that does not change the store type" recipe](https://github.com/pmndrs/zustand/blob/main/docs/typescript.md#middleware-that-does-not-change-the-store-type) and #710. But at the same time notice how the implementation ie `metaStateImpl` is much cleaner and type-free, it requires no type annotations or assertions. And it'll also catch if you make mistakes eg: https://tsplay.dev/WvA8rm

Also looking at your error message, I wonder if you're using v3, note that this answer is for v4.
